### PR TITLE
PaintTrack Start

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Not sure how far I will take this. For now this is just to try understand how its all being done.

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
Must implement `0x004A0AE5` before this can be merged.

**Review by LeftofZen — APPROVED** [review]
Playtested this branch for a while and all the rendering looked fine, I couldn't see any issues on my end

**Comment by AaronVanGeffen** [comment]
> Playtested this branch for a while and all the rendering looked fine, I couldn't see any issues on my end

Is this ready for merging? I was under the impression it wasn't, because of all the `writeRet`s.

**Comment by duncanspumpkin** [comment]
> > Playtested this branch for a while and all the rendering looked fine, I couldn't see any issues on my end
> 
> Is this ready for merging? I was under the impression it wasn't, because of all the `writeRet`s.

The rets will be kept in but i think there is still something i thought was missing.

File: src/OpenLoco/Paint/PaintTrack.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        }
+        const auto height = elTrack.baseZ() * 4;
+        const auto rotation = (session.getRotation() + elTrack.unkDirection()) & 0x3;
+        if (session.getViewFlags() & (1 << 3) && session.getContext()->zoom_level == 0)
+        {
+            const bool isRight = elTrack.isFlag6();
+            if (elTrack.sequenceIndex() == 0 || isRight)
+            {
+                session.setItemType(Ui::ViewportInteraction::InteractionItem::noInteraction);
+                // const auto imageId = Gfx::recolour(getHeightMarkerImage(isRight, elTrack.trackId(), height), Colour::blue);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Paint/PaintTrack.cpp:48]: This bit is missing.

**Comment by duncanspumpkin** [comment]
Okay i went in and added the missing code.

File: src/OpenLoco/Paint/PaintTrack.cpp [hunk 2]
```diff
@@ -39,17 +40,19 @@ namespace OpenLoco::Paint
...
                 session.setItemType(Ui::ViewportInteraction::InteractionItem::noInteraction);
-                // const auto imageId = Gfx::recolour(getHeightMarkerImage(isRight, elTrack.trackId(), height), Colour::blue);
-                // const Map::Pos3 offset(16, 16, height + getHeightMarkerOffset(isRight, elTrack.trackId()) + 8);
-                // const Map::Pos3 bbOffset(1000, 1000, 1087);
-                // const Map::Pos3 bbSize(1, 1, 0);
-                // session.addToPlotListAsParent(imageId, offset, bbOffset, bbSize);
+                const auto markerHeight = height + getTrackDecorationHeightOffset(isFirstTile, elTrack.trackId()) + 8;
+                const auto imageId = ImageId{ getHeightMarkerImage(markerHeight), Colour::blue };
+                const Map::Pos3 offset(16, 16, markerHeight);
+                const Map::Pos3 bbOffset(1000, 1000, 1087);
```
Comment (LeftofZen) [inline src/OpenLoco/Paint/PaintTrack.cpp:53]: what are these offsets for?

File: src/OpenLoco/Graphics/ImageIds.h [hunk 3]
```diff
@@ -1015,6 +1015,8 @@ namespace OpenLoco::ImageIds
     constexpr uint32_t music_controls_next = 2385;
     constexpr uint32_t refit_cargo_button = 2386;
     constexpr uint32_t tab_vehicle_background = 2387;
+    constexpr uint32_t tab = 2387;
```
Comment (LeftofZen) [inline src/OpenLoco/Graphics/ImageIds.h:1018]: this has the same number as the above `tab_vehicle_background`?

File: src/OpenLoco/Paint/PaintTrack.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Paint/PaintTrack.cpp:53]: By having a ridiculously large boundbox offset it forces the height label to be displayed infront of everything

File: src/OpenLoco/Graphics/ImageIds.h (hunk 3 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Graphics/ImageIds.h:1018]: Yeah need to look into this. I just noticed it out of place. I think tab_vehicle can go.

File: src/OpenLoco/Paint/PaintTrack.cpp [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+            regs.ebp = elTrack.sequenceIndex();
+            regs.ecx = rotation;
+            regs.dx = height;
+            call(trackExtraPaintFunc, regs);
+        }
+    }
+
+    void registerTrackHooks()
+    {
+        writeRet(0x004125D8);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Paint/PaintTrack.cpp:123]: Could you add a comment indicating what the purpose of these `writeRet`s is?

**Review by AaronVanGeffen — APPROVED** [review]
Haven't encountered any obvious drawing problems in tracks or roads on this PR. Looks like you've laid some good fundamentals for future implementation.

**Comment by AaronVanGeffen** [comment]
This PR appears to have caused a divergence:
```
$ git bisect log
git bisect start
# status: waiting for both good and bad commits
# good: [2e3bbc55a044bcf9ab8cc751c57defe923aa6b25] Merge pull request #1580 from janisozaur/hook-return-types
git bisect good 2e3bbc55a044bcf9ab8cc751c57defe923aa6b25
# status: waiting for bad commit, 1 good commit known
# bad: [41dcfebe465d4abb2b6e63bcdaa61a28c216968f] Add <array> include for gcc 12 (#1602)
git bisect bad 41dcfebe465d4abb2b6e63bcdaa61a28c216968f
# good: [5cf6baa39f4f99be6e7a48fa5e815384a74be0b7] Split off Object Index from Object Manager (#1591)
git bisect good 5cf6baa39f4f99be6e7a48fa5e815384a74be0b7
# good: [d429d9c9125cfe9a48ac79d849f672822ad4b9bf] Hook malloc and free a bit higher up in the call stack. Fixes running on Apple M1. (#1600)
git bisect good d429d9c9125cfe9a48ac79d849f672822ad4b9bf
# bad: [21fd83fc3ffc96ea03cd5dbcbe45b9e904f976d5] PaintTrack Start (#1386)
git bisect bad 21fd83fc3ffc96ea03cd5dbcbe45b9e904f976d5
# good: [a7f219f82a866e2b24ded2d7106f3f3b730c8877] Implement AnimationManager::update (#1569)
git bisect good a7f219f82a866e2b24ded2d7106f3f3b730c8877
# first bad commit: [21fd83fc3ffc96ea03cd5dbcbe45b9e904f976d5] PaintTrack Start (#1386)
```

**Pull request merged**